File: 2017/cpp/day17.cpp
#include <iostream>

constexpr int STEP = 367;

struct Node {
  int val;
  Node* next;
};

void print_list(const Node* head) {
  const Node* node = head;
  do {
    std::cout << node->val << "  ";
    node = node->next;
  } while (node != head);
  std::cout << '\n';
}

int main() {
  Node* head = new Node { 0, nullptr };
  head->next = head;

  Node* current = head;

  for (int i=1; i<=2017; i++) {
    // skip STEP steps
    for (int j=0; j<STEP; j++) {
      current = current->next;
    }

    // insert new value
    Node* node = new Node { i, current->next };
    current->next = node;

    // update current
    current = current->next;
  }

  std::cout << "Part 1: " << current->next->val << '\n';

  for (int i=2018; i<=50000000; i++) {
    // skip STEP steps
    for (int j=0; j<STEP; j++) {
      current = current->next;
    }

    // insert new value
    Node* node = new Node { i, current->next };
    current->next = node;

    // update current
    current = current->next;

    if (i % 100000 == 0) {
      std::cout << i << '\n';
    }
  }

  std::cout << "Part 2: " << head->next->val << '\n';
}

File: 2017/rust/src/bin/day17.rs
struct Buffer {
    data: Vec<u32>,
    current: usize,
}

impl Buffer {
    fn skip(&mut self, count: usize) {
        self.current = (self.current + count) % self.data.len();
    }

    fn insert(&mut self, value: u32) {
        self.data.insert(self.current + 1, value);
        self.current += 1;
    }
}

fn main() {
    let step = 367;

    let mut buffer = Buffer {
        data: vec![0],
        current: 0,
    };

    for i in 1..=2017 {
        buffer.skip(step);
        buffer.insert(i);
    }

    println!("Part 1: {}", buffer.data[buffer.current+1]);

    let mut current = 0;
    let mut res = 0;

    for i in 1..=50000000 {
        current = (current + step) % i + 1;

        if current == 1 {
            res = i;
        }
    }

    println!("Part 2: {}", res);
}
